### PR TITLE
feat(7181): provide slicing of CursorValues

### PR DESCRIPTION
**Pull request by wiedld**
## Which issue does this PR close?

Part of #7181  .

## Rationale for this change
In the cascaded merge tree, partially sorted record batches are yielded per merge node. 
Therefore, we need the ability to yield a partial of these representative values (a.k.a. the `CursorValues`).

## What changes are included in this PR?
Slicing the `CursorValues`.
Add tests to demonstrate use with `Cursor::new(sliced_cursor_values)`.
Have test coverage include different data types.

## Are these changes tested?
Yes, tests are added.

## Are there any user-facing changes?
No. `CursorValues` are currently an internal construct.

### CODE REVIEW

File: datafusion/physical-plan/src/sorts/stream.rs
Comment (wiedld) [inline datafusion/physical-plan/src/sorts/stream.rs:130]: The memory reservation is still grown on the line above.
Since the `RowValues` take Arc::rows, it's not tied to the memory reservation. (Meaning, on the RowValue slicing is not adding to the reservation.)

**Comment by alamb** [comment]
I plan to review this tomorrow morning

**Comment by wiedld** [comment]
For any questions on the big picture design, referred to the [diagrams here](https://github.com/wiedld/arrow-datafusion/pull/1).

Note that^^ is a drafted followup PR. 
I've tried to incorporate some of the naming/wordage changes suggested, but other names have not changed yet (since I'm not sure how the design diagrams may impact the recommended wordage).

File: datafusion/physical-plan/src/sorts/cursor.rs
Comment (alamb) [inline datafusion/physical-plan/src/sorts/cursor.rs:141]:
```suggestion
    /// Lower bound within `rows`.
```

File: datafusion/physical-plan/src/sorts/cursor.rs
Comment (alamb) [inline datafusion/physical-plan/src/sorts/cursor.rs:143]:
```suggestion
    /// Upper bound within `rows` (not inclusive)
```

File: datafusion/physical-plan/src/sorts/cursor.rs
Comment (alamb) [inline datafusion/physical-plan/src/sorts/cursor.rs:153]: I think it takes a owned `Rows`, and the `Arc` is created internally

```suggestion
    /// Create a new [`RowValues`] from `Rows`.
```

File: datafusion/physical-plan/src/sorts/cursor.rs
Comment (alamb) [inline datafusion/physical-plan/src/sorts/cursor.rs:45]:
```suggestion
    /// Returns a zero-copy slice of this `CursorValues` with the indicated offset and length.
```

File: datafusion/physical-plan/src/sorts/cursor.rs
Comment (alamb) [inline datafusion/physical-plan/src/sorts/cursor.rs:49]: I didn't see any code that would panic about memory being insufficient -- since this is a zero copy slice, I am not sure the memory check is needed

**Comment by tustvold** [comment]
I wonder if you've thought about storing the slice in the BatchCursor, as opposed to pushing it down to this level. That way the logic would only exist once, as opposed to once for every CursorValues? https://github.com/apache/arrow-datafusion/pull/7842/files#diff-1eb407bf7bf6bcc6115e66c6a0197bba129c90328b9d6cd00a5e1eb5e74daa85R103

It could be a stupid idea though...

File: datafusion/physical-plan/src/sorts/cursor.rs
Comment (alamb) [inline datafusion/physical-plan/src/sorts/cursor.rs:614]: What is the reason for also testing with `nulls_first` / `nulls_last`? I didn't see that the semantics of `slice` differ based on the that setting and thus I don't think these tests add much/any new coverage

I suggest either making a loop in `test_slice_primitive` that uses the different `SortOptions` ensuring they produce the same result or just removing these two tests

File: datafusion/physical-plan/src/sorts/cursor.rs
Comment (alamb) [inline datafusion/physical-plan/src/sorts/cursor.rs:818]: thank you for these comments 👨‍🍳 👌

**Review by alamb — APPROVED** [review]
Thank you @wiedld  -- I think this looks really nice. It is both well documented as well as very thoroughly tested. 

I have some suggestions on improvements to the documentation and tests that I think we should consider prior to merging, but I don't think they are strictly necessary which is why I am approving now.

Thanks again

**Comment by tustvold** [comment]
We should probably confirm this doesn't regress the current sort performance, e.g. by introducing an additional memory indirection when comparing cursors, but otherwise if @alamb is happy lets just move forwards with this

File: datafusion/physical-plan/src/sorts/cursor.rs
Comment (wiedld) [reply datafusion/physical-plan/src/sorts/cursor.rs:49]: Leftover from an earlier iteration. My error. TY for catching.

**Comment by wiedld** [comment]
Run with the application of this slicing code (a single composable merge node) [branch here](https://github.com/wiedld/arrow-datafusion/pull/1).

gcp c3d-standard-8-lssd debian-11 

Screenshot 2023-10-27 at 8 14 50 PM

**Comment by alamb** [comment]
I think we are waiting on some more measurement confirmation that this PR doesn't regress performance. Marking it as draft as we do that so we don't accidentally merge this

**Comment by wiedld** [comment]
I think we should close this @alamb , since it's not a priority at the moment. And whenever we circle back, there will be a very large diff (due to changes). Ok to close?

**Comment by alamb** [comment]
Always good with me to close PRs (we can always reopen them later if needed)

Thanks @wiedld

**Pull request closed without merging**
